EventSourcing.Core/AggregateService.cs

```diff
@@ -46,16 +48,56 @@ public AggregateService(IEventStore<TBaseEvent> store)
       
       return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(aggregateId, events, cancellationToken);
     }
+    
+    private async Task<TAggregate> RehydrateFromSnapshotAsync<TAggregate>(Guid aggregateId,
+      CancellationToken cancellationToken = default) where TAggregate : Aggregate<TBaseEvent>, new()
+    {
+      var latestSnapshot = _store.Snapshots
+        .Where(x => x.AggregateId == aggregateId)
+        .OrderBy(x => x.AggregateVersion)
+        .LastOrDefault();
+      
+      var events = _store.Events
+        .Where(x => x.AggregateId == aggregateId);
+
+      if (latestSnapshot != null)
+        events = events
+          .Where(x => x.AggregateVersion >= latestSnapshot.AggregateVersion)
+          .Prepend(latestSnapshot);
+
+      var orderedEvents = events.OrderBy(x => x.AggregateVersion).ToAsyncEnumerable();
+      return await Aggregate<TBaseEvent>.RehydrateAsync<TAggregate>(aggregateId, orderedEvents, cancellationToken);
+    }
 
     public async Task<TAggregate> PersistAsync<TAggregate>(TAggregate aggregate,
       CancellationToken cancellationToken = default) where TAggregate : Aggregate<TBaseEvent>, new()
     {
       if (aggregate.Id == Guid.Empty)
         throw new ArgumentException("Aggregate.Id cannot be empty", nameof(aggregate));
+
+      var previousVersion = aggregate.UncommittedEvents.FirstOrDefault()?.AggregateVersion ?? 0;
       
       await _store.AddAsync(aggregate.UncommittedEvents.ToList(), cancellationToken);
       aggregate.ClearUncommittedEvents();
+      
+      if (aggregate is ISnapshottable s && s.IntervalExceeded(previousVersion ,aggregate.Version))
```

Comment:
I would perhaps refactor this method to accept ```Aggregate``` as parameter and calculate the ```previousVersion``` inside the method itself.